src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,208 @@
+
+    // All messages that are sent expect a response, except for a StatusReport which would indicate an error and
+    // the end of the transfer.
+    if (!msgTypeData.HasMessageType(Protocols::SecureChannel::MsgType::StatusReport))
+    {
+        sendFlags.Set(Messaging::SendMessageFlags::kExpectResponse);
+    }
+
+    Messaging::ExchangeContext * ec = mExchange.Get();
+
+    // Set the response timeout on the exchange before sending the message.
+    ec->SetResponseTimeout(mTimeout);
+    return ec->SendMessage(msgTypeData.ProtocolId, msgTypeData.MessageType, std::move(msgBuf), sendFlags);
+}
+
+CHIP_ERROR AsyncTransferFacilitator::OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                                       System::PacketBufferHandle && payload)
+{
+    VerifyOrReturnError(mExchange, CHIP_ERROR_INCORRECT_STATE);
+
+    VerifyOrReturnError(ec == mExchange.Get(), CHIP_ERROR_INCORRECT_STATE);
+
+    CHIP_ERROR err =
+        mTransfer.HandleMessageReceived(payloadHeader, std::move(payload), System::SystemClock().GetMonotonicTimestamp());
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(BDX, "OnMessageReceived: Failed to handle message: %" CHIP_ERROR_FORMAT, err.Format());
+
+        // This should notify the tranfer object to abort transfer so it can send a status report across the exchange
+        // when we call ProcessOutputEvents below.
+        mTransfer.AbortTransfer(AsyncResponder::GetBdxStatusCodeFromChipError(err));
+    }
+    else if (!payloadHeader.HasMessageType(MessageType::BlockAckEOF))
+    {
+
+        // Almost every BDX message expect BlockAckEOF will follow up with a response on the exchange.
+        ec->WillSendMessage();
+    }
+
+    ProcessOutputEvents();
+    return err;
+}
+
+void AsyncTransferFacilitator::OnResponseTimeout(Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "OnResponseTimeout, ec: " ChipLogFormatExchange, ChipLogValueExchange(ec));
+    DestroySelf();
+}
+
+CHIP_ERROR AsyncResponder::Init(System::Layer * layer, Messaging::ExchangeContext * exchangeCtx, TransferRole role,
+                                BitFlags<TransferControlFlags> xferControlOpts, uint16_t maxBlockSize,
+                                System::Clock::Timeout timeout)
+{
+    AsyncTransferFacilitator::Init(layer, exchangeCtx, timeout);
+    ReturnErrorOnFailure(mTransfer.WaitForTransfer(role, xferControlOpts, maxBlockSize, timeout));
+    return CHIP_NO_ERROR;
+}
+
+void AsyncResponder::NotifyEventHandled(const TransferSession::OutputEvent & event, CHIP_ERROR status)
+{
+    ChipLogDetail(BDX, "NotifyEventHandled : Event %s Error %" CHIP_ERROR_FORMAT,
+                  TransferSession::OutputEvent::TypeToString(event.EventType), status.Format());
+
+    // If it's a message indicating either the end of the transfer or a timeout reported by the transfer session
+    // or an error occured, we need to call DestroySelf().
```

Comment:
So I think what really needs to be documented here is why we need to call ProcessOutputEvents() at all in this case, instead of just destroying ourselves directly. If that's important, please make sure it's written down _why_ it's important so someone doesn't accidentally break things later.

Reply:
done.